Implement audit planner election results entry card

arsalansufi opened the pull request:
_Commit by commit reviewing recommended_

## Overview

Issue link: https://github.com/votingworks/arlo/issues/1440

The first half of the audit planner UI! A card/form for entry of election results. Below is a haphazard screen capture that captures a good number of the possible states the form can be in:

https://user-images.githubusercontent.com/12616928/190314977-01f6be06-12fd-473a-888d-b729d0d9a93f.mov

## Testing

- [x] Added automated tests
- [x] Tested manually

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/IntegerInput.tsx:
```diff
@@ -0,0 +1,32 @@
+      }}
+      placeholder="0"
+      type="number"
+      value={value === 0 ? '' : `${value}`}
```
Distinguishing an explicitly entered `0` from no input is an interesting edge case. Properly handling this would require making the `value` prop optional, e.g. `value?: number`, and updating consumers / consumer state to distinguish `0` and `undefined` - the latter piece would add some complexity that didn't feel worth it for the rare (I think) case that someone needs to enter 0 votes for a candidate

Right now, if you try entering an explicit `0`, the input will ignore your key press and continue displaying the placeholder `0`. The `0` will still be properly recorded under the hood

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/AuditPlanner.tsx:
```diff
@@ -0,0 +1,52 @@
+      toast.error(err instanceof Error ? err.message : 'Unknown error')
+      return
+    }
+    setElectionResultsEditable(false)
```
In my next PR, I'll continue building out the `planAudit` handler to actually compute sample sizes

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/AuditPlanner.tsx:
```diff
@@ -0,0 +1,52 @@
+        enableEditing={enableEditing}
+        planAudit={planAudit}
+        setElectionResults={setElectionResults}
+      />
```
Likewise, in my next PR, I'll add an `AuditPlanCard` here

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/ElectionResultsCard.tsx:
```diff
@@ -0,0 +1,295 @@
+`
+
+const InnerContainer = styled.div`
+  padding: 24px;
```
I decided to punt on using `rem` for now, despite my earlier mentions of doing so in Slack, since the root `html` font size has been modified from its default `16px`, something that isn't recommended:

https://github.com/votingworks/arlo/blob/84b5494432ce94bc015c71ba0020c0bde7a72f99/client/src/App.css#L53-L56

I'd first want to remove that modification, update the few existing `rem` styles, and then start using `rem`

arsalansufi replied:
Issue created ✅ https://github.com/votingworks/arlo/issues/1594

jonahkagan commented on client/src/components/PublicPages/AuditPlanner/ElectionResultsCard.tsx:
```diff
@@ -0,0 +1,295 @@
+          </AdditionalInputsRow>
+        </InnerContainer>
+
+        <CardActionsRow>
```
Now that I'm seeing this button bar implemented, the buttons feel way too big lol. I'd like to try out some other options

arsalansufi replied:
Switched to our more standard right-aligned buttons (see the updated screen capture in the PR description)

jonahkagan replied:
Nice, those look better

jonahkagan reviewed (commented):
Functionality generally looks great! I've got a bunch of feedback mainly on form design (what you built is pretty true to what I mocked, but now I'm noticing new details). Let me know if it's clear or if you wanna discuss

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/ElectionResultsCard.tsx:
```diff
@@ -15,16 +15,7 @@ import { StyledTable } from '../../Atoms/Table'
 import { sum } from '../../../utils/number'
```
This commit is gonna be easier to review with [`?w=1`](https://github.com/votingworks/arlo/pull/1592/commits/28264e52b278ba589ff94a015a1636dba3f89dae?w=1)

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/IntegerInput.tsx:
```diff
@@ -0,0 +1,32 @@
+  value: number
+}
+
+const IntegerInput: React.FC<IProps> = ({
```
Made the switch! I was admittedly skeptical at first :smile: (I've found that many form libraries either can't support all form needs, e.g. cross-field validation, or have very complex interfaces), but I think react-hook-form escapes those pitfalls. It was able to do everything I needed to with an interface that's intuitive once you get up to speed on it!

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/AuditPlanner.tsx:
```diff
@@ -0,0 +1,52 @@
+    try {
+      validateElectionResults(electionResults)
+    } catch (err) {
+      toast.error(err instanceof Error ? err.message : 'Unknown error')
```
Switched to inline error messages! Using the combination of `react-hook-form` and `@hookform/error-message` (a plugin that helps parse the fairly complex `errors` object returned by `useForm`)

One challenge does still remain, which you can see in the updated screen capture in the PR description: When inline error messages are displayed, the height of the form increases (which I'm glad you brought up in another comment). We can alleviate this by increasing the height of the table rows in the form, but that too has limitations, e.g. if we still want to vertically center content in the table rows, inputs will "jump around" within their row as error messages appear and disappear

arsalansufi replied:
I'll post some screen captures of alternatives in just a sec

arsalansufi replied:
This screen capture gives a sense of what the "tall rows" solution might look like:

https://user-images.githubusercontent.com/12616928/190232280-c20ca0de-a84d-4901-b68a-5b9c6e553ec7.mov

jonahkagan replied:
It's a bit weird that the "At least 1 candidate must have greater than 0 votes" is displayed on all the votes fields, but given that this is an extreme edge case it's probably not a big deal.

arsalansufi replied:
Good call - updated in https://github.com/votingworks/arlo/pull/1592/commits/ede0dc2843dfcafbcb0c31a6dd51727048b47fb9

arsalansufi replied:
Documenting that we met in person to discuss the bigger design questions in this thread 🗣️

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/AuditPlanner.tsx:
```diff
@@ -0,0 +1,52 @@
+const AuditPlanner: React.FC = () => {
+  const [electionResults, setElectionResults] = useState<IElectionResults>(
+    constructInitialElectionResults()
+  )
```
As mentioned in my response to other comments, made the switch!

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/ElectionResultsCard.tsx:
```diff
@@ -0,0 +1,295 @@
+              {electionResults.candidates.map((candidate, i) => (
+                <tr key={candidate.id}>
+                  <td>
+                    {editable ? (
```
Really good point. I went with the latter option, though with an important modification. I used readonly inputs instead of disabled inputs, since the former are better for accessibility. A screen reader will completely ignore disabled inputs but will still read aloud readonly inputs

Regarding form height, curious to hear your thoughts as they relate to inline error messages: https://github.com/votingworks/arlo/pull/1592#discussion_r971136440

jonahkagan replied:
Awesome, I like the readonly solution! I didn't know about that

arsalansufi commented on client/src/components/PublicPages/AuditPlanner/ElectionResultsCard.tsx:
```diff
@@ -0,0 +1,295 @@
+                            icon="minus"
+                            onClick={() => removeCandidate(i)}
+                          />
+                        )}
```
Moved them to the right end of the row! Also changed the icon to a trash can 🗑️ for added clarity

jonahkagan replied:
Trash can gives me a bit of a "delete forever" vibe vs "remove". Did you check out the icons named "remove" and "delete"? I think using one of those plus `minimal` styling plus `danger` intent might be nice.

arsalansufi commented:
@jonahkagan This is ready for a follow-up look!

arsalansufi commented:
@jonahkagan Updated the screen capture to include my final round of design polish following our in-person chat! Really happy with where we've landed, both from a technical and a design perspective!

jonahkagan commented:
This is looking much improved! I know you're trying to close this out and I want to unblock that, but I do want to make a couple more suggestions. Take them or leave them!

1. I think the inputs in the candidates table are a bit awkwardly far apart still. I think one of the root causes is the ratio of the input height to error text height. I tried bumping up input height from 30px to 40px and removing the td bottom padding and that seemed to help.
2. There's a lot of awkward empty space on the right. The remove button in particular is far away from anything. What if we condense the whole thing? I don't think the fields need to be so long. Here's a quick and dirty cut and paste mock up. The actual width in the mockup is incidental, focus on the concept and alignment of fields.
<img width="1022" alt="Screen Shot 2022-09-14 at 5 58 56 PM" src="https://user-images.githubusercontent.com/530106/190289875-618b6030-145c-4c69-867f-80f1b30dd955.png">

jonahkagan commented on client/src/App.css:
```diff
@@ -29,10 +29,6 @@ legend {
 
-:focus {
-  outline: rgb(77, 144, 254) dashed 0.25rem;
-}
```
Hurrah!

jonahkagan commented on client/src/components/PublicPages/AuditPlanner/ElectionResultsCard.tsx:
```diff
@@ -0,0 +1,368 @@
+                        },
+                        valueAsNumber: true,
+                      })}
+                      type="number"
```
Is it possible for these number inputs to start empty instead of starting with 0? 0 doesn't seem like a useful default value since users will almost always need to change it.

arsalansufi replied:
Good call - taken care of in https://github.com/votingworks/arlo/pull/1592/commits/1ec00130e75acae0001724b88433841fb2ea943e

Required tweaking types to allow `null` as a value for numeric inputs. While react-hook-form automatically allows all values to be `undefined`, I had to add and use `null` to get clearing of inputs via the "Clear" button to work

arsalansufi commented:
> I know you're trying to close this out and I want to unblock that, but I do want to make a couple more suggestions. Take them or leave them!

@jonahkagan No worries at all! I really appreciate and enjoy the design feedback!

Incorporated both suggestions 1 (https://github.com/votingworks/arlo/pull/1592/commits/e976e0d3165b1289de348743675f1adbee9a42d5) and 2 (https://github.com/votingworks/arlo/pull/1592/commits/7ff512ac1865846a549baa5d4699f4ef06a7c03a) from your comment and agree that they were the right calls.

Updated the screen capture in the PR description to reflect the latest and greatest. One additional change worth noting: I centered the top-level "Audit Planner" header since the election results entry card is no longer full width
